File: backend/app/schemas/notification.py
"""Pydantic schemas for the notification center + per-user preferences."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.api.pagination import PageMeta
from app.models.notification import NOTIFICATION_EVENT_TYPES


class NotificationResponse(BaseModel):
    id: uuid.UUID
    event_type: str
    entity_type: str
    entity_id: uuid.UUID | None = None
    title: str
    body: str | None = None
    read_at: datetime | None = None
    created_at: datetime


class NotificationListResponse(PageMeta):
    """Paginated list of the current user's notifications.

    `unread` is the total unread count for the user (independent of the page
    window), so the sidebar badge can use the same response.
    """

    items: list[NotificationResponse]
    total: int
    unread: int


class ChannelPrefs(BaseModel):
    """Per-event delivery channels."""

    email: bool = True
    in_app: bool = True


class NotificationPrefs(BaseModel):
    """The full preference map: one ChannelPrefs per notifiable event type.

    **Must cover every entry in `NOTIFICATION_EVENT_TYPES`.** A key missing here
    isn't inert: `notification_dispatch.resolve_prefs` defaults an unknown event
    to **on**, so an event the schema doesn't enumerate is one the user cannot
    turn off. That is how `chat_message` came to email the AP team on every
    supplier-portal message with no opt-out, while `contract_renewal_due` and
    `cash_shortfall_projected` were equally unmutable.

    Drift guard: `tests/test_notification_prefs_roster.py`.
    """

    invoice_assigned: ChannelPrefs = Field(default_factory=ChannelPrefs)
    invoice_approved: ChannelPrefs = Field(default_factory=ChannelPrefs)
    invoice_rejected: ChannelPrefs = Field(default_factory=ChannelPrefs)
    invoice_paid: ChannelPrefs = Field(default_factory=ChannelPrefs)
    contract_renewal_due: ChannelPrefs = Field(default_factory=ChannelPrefs)
    chat_message: ChannelPrefs = Field(default_factory=ChannelPrefs)
    cash_shortfall_projected: ChannelPrefs = Field(default_factory=ChannelPrefs)


class NotificationPrefsUpdate(BaseModel):
    """Partial update — only the supplied event types are changed.

    Same roster obligation as `NotificationPrefs`: an event type absent here is
    one a user can read but never change.
    """

    invoice_assigned: ChannelPrefs | None = None
    invoice_approved: ChannelPrefs | None = None
    invoice_rejected: ChannelPrefs | None = None
    invoice_paid: ChannelPrefs | None = None
    contract_renewal_due: ChannelPrefs | None = None
    chat_message: ChannelPrefs | None = None
    cash_shortfall_projected: ChannelPrefs | None = None


class DeviceTokenRegister(BaseModel):
    """`POST /api/notifications/device-token` body — registers (or
    re-registers) the caller's current push token for one platform.

    Registration only. There is no server-side push-SENDING path yet (no
    Firebase Admin SDK adapter exists in this codebase) — this just persists
    the token for whenever that's built. At most one token per platform per
    user: a fresh registration replaces whatever was stored, since FCM tokens
    rotate and a stale one should be overwritten, not accumulated.
    """

    token: str = Field(min_length=1, max_length=4096)
    platform: Literal["ios", "android"]


class DeviceTokenResponse(BaseModel):
    platform: Literal["ios", "android"]
    updated_at: datetime


class MarkReadResponse(BaseModel):
    id: uuid.UUID
    read_at: datetime | None = None


class ReadAllResponse(BaseModel):
    updated: int


class UnreadCountResponse(BaseModel):
    unread: int


# Re-export for callers that want the canonical event list.
__all__ = [
    "NotificationResponse",
    "NotificationListResponse",
    "ChannelPrefs",
    "NotificationPrefs",
    "NotificationPrefsUpdate",
    "DeviceTokenRegister",
    "DeviceTokenResponse",
    "MarkReadResponse",
    "ReadAllResponse",
    "UnreadCountResponse",
    "NOTIFICATION_EVENT_TYPES",
]

File: frontend/src/lib/types/notification.ts
import type { MessageKey } from '$lib/i18n/messages';

/**
 * The notifiable event vocabulary.
 *
 * **The server owns this list.** The canonical roster is
 * `backend/app/models/notification.py::NOTIFICATION_EVENT_TYPES`; this union —
 * and `EVENT_ORDER` below — must carry the same strings *in the same order*.
 * `notification.roster.test.ts` reads that Python file as text and fails if the
 * two drift, because the failure mode is silent: `notification_dispatch
 * .resolve_prefs` defaults a *missing* preference key to **on**, so an event
 * the frontend never lists is an event the user cannot switch off. Three of
 * these (`contract_renewal_due`, `chat_message`, `cash_shortfall_projected`)
 * shipped server-side while this union still stopped at the four `invoice_*`
 * ones — `chat_message` emails the AP team on every supplier-portal message,
 * with no way to mute it from the UI.
 */
export type NotificationEventType =
	| 'invoice_assigned'
	| 'invoice_approved'
	| 'invoice_rejected'
	| 'invoice_paid'
	| 'contract_renewal_due'
	| 'chat_message'
	| 'cash_shortfall_projected';

export interface Notification {
	id: string;
	event_type: NotificationEventType;
	entity_type: string;
	entity_id: string | null;
	title: string;
	body: string | null;
	read_at: string | null;
	created_at: string;
}

export interface NotificationListResponse {
	items: Notification[];
	total: number;
	unread: number;
	page: number;
	page_size: number;
}

export interface ChannelPrefs {
	email: boolean;
	in_app: boolean;
}

/** The complete preference map — one {email, in_app} pair per event type. */
export type NotificationPrefs = Record<NotificationEventType, ChannelPrefs>;

/**
 * The server's default for an unset preference: both channels on. Mirrors
 * `notification_dispatch._DEFAULT_CHANNELS` — opt-out, not opt-in.
 */
export const DEFAULT_CHANNEL_PREFS: ChannelPrefs = { email: true, in_app: true };

/**
 * Fill in any event the server omitted, with the server's own default.
 *
 * `GET /api/notifications/preferences` returns whatever
 * `schemas/notification.py::NotificationPrefs` declares — so a backend that has
 * not yet grown a field for a newer event type answers with a *partial* map.
 * Indexing that map for a row the grid renders would hand the template
 * `undefined` and throw on `.in_app`, taking the whole Notifications card down.
 * Normalising instead degrades to "shown as on, the same state the server will
 * actually apply", which is the truth rather than a crash.
 */
export function normalizePrefs(
	raw: Partial<NotificationPrefs> | null | undefined
): NotificationPrefs {
	const out = {} as NotificationPrefs;
	for (const event of EVENT_ORDER) {
		const channels = raw?.[event];
		out[event] = {
			email: channels?.email ?? DEFAULT_CHANNEL_PREFS.email,
			in_app: channels?.in_app ?? DEFAULT_CHANNEL_PREFS.in_app
		};
	}
	return out;
}

/**
 * Display labels for each event type — English, data-driven, the same
 * convention as `chatNotifications.ts`'s `CHAT_EVENT_LABELS`.
 *
 * Used by the notification centre and the bell, neither of which is in the
 * i18n extraction slice yet. The `/profile` preference grid renders
 * `EVENT_LABEL_KEYS` instead; the roster guard pins the English catalogue
 * values to this map so the two spellings can never diverge.
 */
export const EVENT_LABELS: Record<NotificationEventType, string> = {
	invoice_assigned: 'Invoice assigned to me',
	invoice_approved: 'Invoice approved',
	invoice_rejected: 'Invoice rejected',
	invoice_paid: 'Invoice paid',
	contract_renewal_due: 'Contract renewal due',
	chat_message: 'New supplier chat message',
	cash_shortfall_projected: 'Projected cash shortfall'
};

/**
 * The i18n key carrying each event's label, for surfaces that are inside the
 * extraction slice (the `/profile` preference grid). Split from `EVENT_LABELS`
 * rather than replacing it because the bell and centre are not extracted yet —
 * and kept honest by the roster guard, which asserts `en[key] === label`.
 */
export const EVENT_LABEL_KEYS: Record<NotificationEventType, MessageKey> = {
	invoice_assigned: 'profile.notifications.event.invoiceAssigned',
	invoice_approved: 'profile.notifications.event.invoiceApproved',
	invoice_rejected: 'profile.notifications.event.invoiceRejected',
	invoice_paid: 'profile.notifications.event.invoicePaid',
	contract_renewal_due: 'profile.notifications.event.contractRenewalDue',
	chat_message: 'profile.notifications.event.chatMessage',
	cash_shortfall_projected: 'profile.notifications.event.cashShortfallProjected'
};

/**
 * Render order for the preference grid. Deliberately the backend's own
 * declaration order in `NOTIFICATION_EVENT_TYPES` — invoice lifecycle first,
 * then the sweep-driven events in the order they were added — so there is one
 * ordering to reason about instead of a frontend opinion that silently drifts.
 */
export const EVENT_ORDER: NotificationEventType[] = [
	'invoice_assigned',
	'invoice_approved',
	'invoice_rejected',
	'invoice_paid',
	'contract_renewal_due',
	'chat_message',
	'cash_shortfall_projected'
];
